### PR TITLE
Replace simple_hasher::Hasher with AlgebraicHasher

### PR DESCRIPTION
**Pull request by sshine**
The AlgebraicHasher trait outputs a `Digest`, consisting of `DIGEST_LEN` `BFieldElement`s, no matter if `blake3::Hasher` or `RescuePrimeRegular` is used. This means that `blake3::Hasher` can be used throughout the pipeline, including in Triton VM where digests have an expected format and size.

In `simple_hasher`:
- Remove `Hasher`
- Remove `SamplableFrom` in favor of `XFieldElement::sample()`
- Move `Hashable` to `AlgebraicHasher`

Associated types and type parameters are hardcoded in `AlgebraicHasher`
- `Hashable`'s parameter is always `BFieldElement`
- `Hasher`'s `T` (the associated type denoting the unit element of a digest) is always `BFieldElement`
- `Hasher`'s `Digest` (the associated type) is always `Digest` (the struct)

Refactor `.emojihash()`:
- Drop `.emojihash_raw()`
- Use `u64::to_be_bytes()` instead of Hasher's `.to_sequence()`

`impl Distribution<Digest> for Standard`:
- Makes `random_values()` capable of generating random digests.

Add `PhantomData<H>` back to:
- `MerkleTree`
- `MmrAccumulator`

...since they do not pass their 'H' onto their fields' types.

### CODE REVIEW

**Comment by sshine** [comment]
**Note:** The last commit awaits one minor fix before it is reworded.

File: stark-brainfuck/src/stark.rs [hunk 1]
```diff
@@ -593,51 +579,24 @@ impl Stark {
...
-        //         || vec![XFieldElement::zero(); self.fri.domain.length],
-        //         |acc, weighted_terms| {
-        //             acc.iter()
-        //                 .zip(weighted_terms.iter())
-        //                 .map(|(a, wt)| *a + *wt)
-        //                 .collect()
-        //         },
-        //     );
-
         timer.elapsed("combination_codeword");
```
Comment (Ulrik-dk) [inline stark-brainfuck/src/stark.rs:582]: This timer mark is now superfluous.

File: twenty-first/src/shared_math/fri.rs [hunk 2]
```diff
@@ -336,23 +332,22 @@ where
...
 
         for _ in 0..num_rounds {
             // Get a challenge from the proof stream
-            let alpha: XFieldElement =
-                XFieldElement::from_vecu8(proof_stream.verifier_fiat_shamir());
+            let challenge: Digest = proof_stream.verifier_fiat_shamir();
+            let alpha: XFieldElement = XFieldElement::sample(&challenge);
             alphas.push(alpha);
-            roots.push(proof_stream.dequeue(32)?);
+            roots.push(proof_stream.dequeue(Digest::BYTES)?);
```
Comment (Ulrik-dk) [inline twenty-first/src/shared_math/fri.rs:350]: Thanks for getting rid of some of these magical constants.

File: twenty-first/src/util_types/mmr/mmr_membership_proof.rs [hunk 3]
```diff
@@ -1316,55 +1309,55 @@ mod mmr_membership_proof_test {
...
-                .map(|x| rp.hash_sequence(&vec![BFieldElement::new(x as u64)]))
-                .collect_vec();
+        for leaf_count in 0..9 {
+            let leaf_hashes: Vec<Digest> = random_elements(leaf_count);
             // let archival_mmr = ArchivalMmr::<RescuePrimeRegular>::new(leaf_hashes.clone());
-            let mut archival_mmr: ArchivalMmr<Hasher> =
+            let mut archival_mmr: ArchivalMmr<H> =
                 get_archival_mmr_from_digests(leaf_hashes.clone());
-            let new_leaf = rp.hash_sequence(&vec![BFieldElement::new(13333337)]);
+            let new_leaf = H::hash(&BFieldElement::new(13333337));
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_membership_proof.rs:1322]: Is this a magic number or a random Digest?

File: twenty-first/src/util_types/mmr/mmr_membership_proof.rs [hunk 4]
```diff
@@ -1188,110 +1156,135 @@ mod mmr_membership_proof_test {
...
     fn update_membership_proof_from_append_big_blake3() {
-        type Digest = Blake3Hash;
-        type Hasher = blake3::Hasher;
+        type H = blake3::Hasher;
 
         // Build MMR from leaf count 0 to 68, and loop through *each*
         // leaf index for MMR, modifying its membership proof with an
         // append update.
-        let new_leaf: Digest = 133333333333333333333337u128.into();
+        let new_leaf: Digest = H::hash(&133333333333333333333337u128);
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_membership_proof.rs:1196]: PLACEHOLDER TEXT.

Comment (Ulrik-dk) [reply twenty-first/src/util_types/mmr/mmr_membership_proof.rs:1196]: Is this a magic number or a random digest?

File: twenty-first/src/util_types/mmr/mmr_membership_proof.rs [hunk 5]
```diff
@@ -718,20 +691,19 @@ mod mmr_membership_proof_test {
...
+        let mut membership_proofs: Vec<MmrMembershipProof<H>> = vec![];
+        for i in 0..total_leaf_count {
+            let data_index = i as u128;
             membership_proofs.push(archival_mmr.prove_membership(data_index).0);
         }
 
-        let new_leaf2: Digest = 133337u128.into();
-        let new_leaf3: Digest = 12345678u128.into();
+        let new_leaf2: Digest = H::hash(&133337u128);
+        let new_leaf3: Digest = H::hash(&12345678u128);
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_membership_proof.rs:706]: `new_leaf2` and `new_leaf3`: Is this a magic number or a random digest?

File: twenty-first/src/util_types/mmr/mmr_accumulator.rs [hunk 6]
```diff
@@ -603,38 +557,37 @@ mod accumulator_mmr_tests {
...
-                get_archival_mmr_from_digests(bad_digests.clone());
-            let bad_membership_proof: MmrMembershipProof<Hasher> = bad_mmr.prove_membership(0).0;
+
+            let mut bad_mmr: ArchivalMmr<H> = get_archival_mmr_from_digests(bad_digests.clone());
+            let bad_membership_proof: MmrMembershipProof<H> = bad_mmr.prove_membership(0).0;
             let bad_membership_proof_digest = bad_digests[0];
-            let bad_leaf: Digest = 8765432165123u128.into();
-            // let archival_mmr_init = ArchivalMmr::<Hasher>::new(leaf_hashes_start.clone());
-            let mut archival_mmr_init: ArchivalMmr<Hasher> =
+            let bad_leaf: Digest = H::hash(&8765432165123u128);
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_accumulator.rs:572]: Is this a magic number or a random digest?

File: twenty-first/src/util_types/mmr/mmr_accumulator.rs [hunk 7]
```diff
@@ -603,38 +557,37 @@ mod accumulator_mmr_tests {
...
+            let mut archival_mmr_init: ArchivalMmr<H> =
                 get_archival_mmr_from_digests(leaf_hashes_start.clone());
-            let mut accumulator_mmr = MmrAccumulator::<Hasher>::new(leaf_hashes_start.clone());
+            let mut accumulator_mmr = MmrAccumulator::<H>::new(leaf_hashes_start.clone());
+
             for append_size in 0..18 {
-                let appends: Vec<Digest> = (2000u128..2000u128 + append_size)
-                    .map(|x| x.into())
+                let appends: Vec<Digest> = (2000..2000 + append_size)
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_accumulator.rs:578]: Are these magic numbers?

File: twenty-first/src/util_types/mmr/mmr_membership_proof.rs [hunk 8]
```diff
@@ -675,14 +648,14 @@ mod mmr_membership_proof_test {
 
         // Increase size to 8 and verify that the peaks are now different
         mmr_size = 8;
-        let leaf_hash: Digest = 1337u128.into();
+        let leaf_hash: Digest = H::hash(&1337u128);
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_membership_proof.rs:651]: Is this a magic number or a random digest?

File: twenty-first/src/util_types/mmr/mmr_membership_proof.rs [hunk 9]
```diff
@@ -691,7 +664,7 @@ mod mmr_membership_proof_test {
 
         // Increase size to 9 and verify that the peaks are now different
         mmr_size = 9;
-        let another_leaf_hash: Digest = 13337u128.into();
+        let another_leaf_hash: Digest = H::hash(&13337u128);
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_membership_proof.rs:667]: Magic number or random digest?

File: twenty-first/src/util_types/mmr/mmr_membership_proof.rs [hunk 10]
```diff
@@ -990,19 +957,17 @@ mod mmr_membership_proof_test {
...
     fn update_membership_proof_from_leaf_mutation_test() {
-        type Digest = Blake3Hash;
-        type Hasher = blake3::Hasher;
+        type H = blake3::Hasher;
 
-        let leaf_hashes: Vec<Digest> = (14u128..14 + 8).map(|x| x.into()).collect();
-        let new_leaf: Digest = 133337u128.into();
-        let mut accumulator_mmr = MmrAccumulator::<Hasher>::new(leaf_hashes.clone());
+        let leaf_hashes: Vec<Digest> = random_elements(8);
+        let new_leaf: Digest = H::hash(&133337u128);
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_membership_proof.rs:963]: Is this a magic number or a random digest?

File: twenty-first/src/util_types/mmr/mmr_membership_proof.rs [hunk 11]
```diff
@@ -1107,57 +1072,61 @@ mod mmr_membership_proof_test {
...
         // Build MMR from leaf count 0 to 22, and loop through *each*
         // leaf index for MMR, modifying its membership proof with a
         // leaf update.
         for leaf_count in 0..=22 {
-            let start: u128 = 543217893265643843678;
-            let leaf_hashes: Vec<Digest> = (start..start + leaf_count).map(|x| x.into()).collect();
-            let new_leaf: Digest = 133333333333333333333337u128.into();
-            let mut archival_mmr: ArchivalMmr<Hasher> =
+            let leaf_hashes: Vec<Digest> = random_elements(leaf_count);
+            let new_leaf: Digest = H::hash(&133333333333333333333337u128);
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_membership_proof.rs:1082]: Is this a magic number or a random digest?

File: twenty-first/src/util_types/mmr/mmr_membership_proof.rs [hunk 12]
```diff
@@ -1167,19 +1136,18 @@ mod mmr_membership_proof_test {
...
-        type Hasher = blake3::Hasher;
+        type H = blake3::Hasher;
 
         let leaf_count = 7;
-        let leaf_hashes: Vec<Digest> = (14u128..14 + leaf_count).map(|x| x.into()).collect();
-        let new_leaf: Digest = 133337u128.into();
-        let mut archival_mmr: ArchivalMmr<Hasher> =
-            get_archival_mmr_from_digests(leaf_hashes.clone());
+        let leaf_hashes: Vec<Digest> = random_elements(leaf_count);
+        let new_leaf: Digest = H::hash(&133337u128);
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_membership_proof.rs:1143]: Is this a magic number or a random digest?

File: twenty-first/src/util_types/mmr/mmr_accumulator.rs [hunk 13]
```diff
@@ -603,38 +557,37 @@ mod accumulator_mmr_tests {
...
+                    .map(|x| H::hash(&(x as u128)))
                     .collect();
                 let mutate_count = cmp::min(12, start_size);
                 for mutate_size in 0..mutate_count {
-                    let new_leaf_values: Vec<Digest> =
-                        (13u128..13u128 + mutate_size).map(|x| x.into()).collect();
+                    let new_leaf_values: Vec<Digest> = (13..13 + mutate_size)
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_accumulator.rs:583]: Are these 13's magic numbers?

File: twenty-first/src/util_types/mmr/mmr_accumulator.rs [hunk 14]
```diff
@@ -603,38 +557,37 @@ mod accumulator_mmr_tests {
...
-        type Digest = Blake3Hash;
-        type Hasher = blake3::Hasher;
+        type H = blake3::Hasher;
 
         for start_size in 1..35 {
-            let leaf_hashes_start: Vec<Digest> = (4000u128..4000u128 + start_size)
-                .map(|x| x.into())
+            let leaf_hashes_start: Vec<Digest> = random_elements(start_size);
+
+            let bad_digests: Vec<Digest> = (12..12 + start_size)
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_accumulator.rs:565]: Magic number?

File: twenty-first/src/util_types/mmr/mmr_accumulator.rs [hunk 15]
```diff
@@ -368,51 +344,43 @@ mod accumulator_mmr_tests {
...
-        let mut archive_mmr_start: ArchivalMmr<Hasher> =
+        type H = blake3::Hasher;
+
+        let leaf_hashes_start: Vec<Digest> =
+            [14, 15, 16, 18].iter().map(|x: &u128| H::hash(x)).collect();
+
+        let new_leaf_value: Digest = H::hash(&17u128);
+
+        let leaf_hashes_end: Vec<Digest> =
+            [14, 15, 16, 17].iter().map(|x: &u128| H::hash(x)).collect();
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/mmr_accumulator.rs:376]: Magic numbers?

File: twenty-first/src/util_types/mmr/archival_mmr.rs [hunk 16]
```diff
@@ -125,23 +122,23 @@ where
...
-    H: Hasher,
-    u128: Hashable<H::T>,
-{
+impl<H: AlgebraicHasher> ArchivalMmr<H> {
     /// Create a new, empty archival MMR
     pub fn new(db: DB) -> Self {
-        let mut db_vector: DatabaseVector<H::Digest> = DatabaseVector::new(db);
-        let dummy_digest = H::new().hash_sequence(&0u128.to_sequence());
+        let mut db_vector: DatabaseVector<Digest> = DatabaseVector::new(db);
+        let dummy_digest = H::hash(&0u128);
```
Comment (Ulrik-dk) [inline twenty-first/src/util_types/mmr/archival_mmr.rs:129]: Magic number or random digest?

File: twenty-first/src/shared_math/rescue_prime_regular.rs [hunk 17]
```diff
@@ -1079,47 +1076,16 @@ impl RescuePrimeRegular {
...
-    }
-}
-
-impl SamplableFrom<[BFieldElement; 5]> for XFieldElement {
-    fn sample(digest: &[BFieldElement; 5]) -> Self {
-        XFieldElement {
-            coefficients: digest[0..3].try_into().unwrap(),
-        }
+    fn hash_pair(left: &Digest, right: &Digest) -> Digest {
+        let mut input = [BFieldElement::zero(); 10];
```
Comment (Ulrik-dk) [inline twenty-first/src/shared_math/rescue_prime_regular.rs:1085]: Maybe use `DIGEST_LENGTH*2` here instead of 10.

**Review by Ulrik-dk — CHANGES_REQUESTED** [review]
Most of my comments are on seemingly magical numbers, which are used in old tests, instead of the new random element function. In other places, genuinely magical numbers that are needed multiple places can be factored out, and referenced in these places.

This issue is many places in the code, and you're not responsible for that. Maybe it should be an issue on its own, to factor out magical numbers. Or to just be more conscious of it.

**Comment by Sword-Smith** [comment]
It's fine to use so-called magic numbers in tests. Having tests with concrete numbers/values is a test on a level of abstraction that's lower than auto generating values for tests. To convince yourself that something is working, you need, in my opinion, both values that are hardcoded where you can reason about the expected output **and** auto-generated values to have a chance of testing the function more thoroughly.

Declaring values used in tests as constant just increases the indirection when trying to decipher why a test i failing.

Let's say you have an `add` function. Then one of the tests could look like this
```rust
fn test_add_simple() {
    assert_eq!(15, add(8, 7));
    assert_eq!(40000, add(8, 39992));
}
```

Declaring the tests values outside of the test itself (as constant in modular scope or just as separate lines inside the test function) gives worse code. If something fails you want to see the values on the line where the test is called. Declaring the values somewhere else prevents you from doing this.

Of course, if the values are re-used and interlinked, I think it makes sense to have a declaration within a test function as

```rust
fn test_add_simple() {
    let expected_sum = 1337;
    for i in 0..=expected_sum {
        assert_eq!(expected_sum, add(expected_sum - i, i));
    }
}
```

**Comment by Sword-Smith** [comment]
Looks reasonable to me. I look forward to getting a `Digest` type in here instead of having it in the client software.

**Review by Ulrik-dk — APPROVED** [review]

**Pull request merged**